srml/contract/src/rent.rs

```diff
@@ -0,0 +1,138 @@
+use crate::{BalanceOf, ContractInfo, ContractInfoOf, Module, TombstoneContractInfo, Trait};
+use runtime_primitives::traits::{As, Bounded, CheckedDiv, CheckedMul, Saturating, Zero};
+use srml_support::traits::{Currency, ExistenceRequirement, Imbalance, WithdrawReason};
+use srml_support::StorageMap;
+
+/// Evict and optionnaly pay rent, at block number. Return if evicted
+///
+/// Exempted from rent iff:
+/// * rent is offset completely by the `rent_deposit_offset`,
+/// * or rent has already been paid for this block number,
+/// * or account doesn't have a contract,
+/// * or account has a tombstone.
+///
+/// Evicted iff:
+/// * rent exceed rent allowance,
+/// * or can't withdraw the rent,
+/// * or go below subsistence threshold.
+fn try_evict_or_and_pay_rent<T: Trait>(
+	account: &T::AccountId,
+	block_number: T::BlockNumber,
+	pay_rent: bool,
+) -> bool {
+	let contract = match <ContractInfoOf<T>>::get(account) {
+		None | Some(ContractInfo::Tombstone(_)) => return false,
+		Some(ContractInfo::Alive(contract)) => contract,
+	};
+
+	// Rent has already been paid
+	if contract.deduct_block >= block_number {
+		return false;
+	}
+
+	let balance = T::Currency::free_balance(account);
+
+	let free_storage = balance
+		.checked_div(&<Module<T>>::rent_deposit_offset())
+		.unwrap_or(<BalanceOf<T>>::sa(0));
+
+	let effective_storage_size =
+		<BalanceOf<T>>::sa(contract.storage_size).saturating_sub(free_storage);
+
+	let fee_per_block = effective_storage_size
+		.checked_mul(&<Module<T>>::rent_byte_price())
+		.unwrap_or(<BalanceOf<T>>::max_value());
+
+	if fee_per_block.is_zero() {
+		// The rent deposit offset reduced the fee to 0. This means that the contract
+		// gets the rent for free.
+		return false;
+	}
+
+	let blocks_to_rent = block_number.saturating_sub(contract.deduct_block);
+	let rent = fee_per_block
```

Comment:
Let's use `dues` terminology to be aligned more with the staterent sim and spec?